exonum/src/events/network.rs

```diff
@@ -369,47 +380,73 @@ impl Listener {
-                    Some(Ok(other)) => Err(other_error(&format!(
-                        "First message is not Connect, got={:?}",
-                        other
-                    ))),
-                    Some(Err(e)) => Err(into_other(e)),
-                    None => Err(other_error("Incoming socket closed")),
-                })
-                .and_then(move |(connect, stream)| {
-                    trace!("Received handshake message={:?}", connect);
-                    let event = NetworkEvent::PeerConnected(addr, connect);
-                    let stream = network_tx
-                        .clone()
-                        .send(event)
-                        .map_err(into_other)
-                        .and_then(move |_| Ok(stream))
-                        .flatten_stream();
-
-                    stream.for_each(move |raw| {
-                        let event = NetworkEvent::MessageReceived(addr, raw);
-                        network_tx.clone().send(event).map_err(into_other).map(drop)
-                    })
-                })
-                .map(|_| {
-                    // Ensure that holder lives until the stream ends.
-                    let _holder = holder;
+            let connection_handler = handshake
+                .listen(sock)
+                .and_then(move |(sock, key)| {
+                    let (_, stream) = sock.split();
+                    stream
+                        .into_future()
+                        .map_err(|e| e.0)
+                        .and_then(|(raw, stream)| (Self::parse_connect_msg(raw), Ok(stream)))
+                        .and_then(move |(connect, stream)| {
+                            trace!("Received handshake message={:?}", connect);
+
+                            Self::process_incoming_messages(
+                                stream,
+                                network_tx,
+                                connect,
+                                ConnectInfo {
+                                    address: addr,
+                                    public_key: key,
+                                },
+                            )
+                        })
+                        .map(|_| {
+                            // Ensure that holder lives until the stream ends.
+                            let _holder = holder;
+                        })
                 })
                 .map_err(log_error);
+
             handle.spawn(to_box(connection_handler));
             to_box(future::ok(()))
         });
 
         Ok(Listener(to_box(server)))
     }
+
+    fn parse_connect_msg(raw: Option<RawMessage>) -> Result<Connect, io::Error> {
+        let raw = raw.ok_or_else(|| other_error("Incoming socket closed"))?;
+        let message = Any::from_raw(raw).map_err(into_other)?;
+        match message {
+            Any::Connect(connect) => Ok(connect),
+            other => Err(other_error(&format!(
+                "First message is not Connect, got={:?}",
```

Comment:
Nit: I think expanding to "First message from a remote peer is not `Connect`" could be more helpful.